pkg/controller/operators/adoption_controller.go

```diff
@@ -512,15 +498,31 @@ func (r *AdoptionReconciler) mapToProviders(obj client.Object) (requests []recon
 	return
 }
 
-func flatten(objs []runtime.Object) (flattened []runtime.Object) {
+// flatten extracts slice of ObjectList to the slice of Object
+func flatten(objs []client.ObjectList) (flattened []client.Object) {
 	for _, obj := range objs {
-		if nested, err := meta.ExtractList(obj); err == nil {
-			flattened = append(flattened, flatten(nested)...)
+		nested, err := meta.ExtractList(obj)
+		if err != nil {
 			continue
 		}
 
-		flattened = append(flattened, obj)
-	}
+		if len(nested) == 0 {
+			continue
+		}
 
+		if meta.IsListType(nested[0]) {
+			cNested := make([]client.ObjectList, len(nested))
+			for index := range nested {
+				cNested[index] = nested[index].(client.ObjectList)
+			}
+			flattened = append(flattened, flatten(cNested)...)
```

Comment:
Is is possible for individual items within an ObjectList to be more ObjectLists? I don't _think_ so, but I could be wrong.

In the previous implementation, we didn't know if the `obj []runtime.Object`s were lists or just regular objects. Hence the need to try to call `meta.ExtractList` (and if that didn't work, just add the object directly).

Now that we know that the objects being passed in are lists, we don't need to handle both cases.

Would this simpler version work instead:
```go
func flatten(lists []client.ObjectList) (flattened []client.Object) {
	for _, list := range lists {
		_ = meta.EachListItem(list, func(obj runtime.Object) error {
			flattened = append(flattened, obj.(client.Object))
			return nil
		})
	}
	return
}
```

Reply:
@joelanford Thank you for the explanation! 

Yes, what you say makes sense to me. I will use the function you provide in the above comment